### DISCUSSION TITLE
When is focus and online related global window listeners removed/cleaned up ?

### DISCUSSION THREAD

**Discussion by omar-oslo** [TanStack/query#4206] [Q&A] [answered]
We are using [react-query 2.26.4](https://github.dev/TanStack/query/tree/v2.26.4) and found that the following event listeners are not cleaned up. 

>   window.addEventListener('visibilitychange', handleFocus, false)
>   window.addEventListener('focus', handleFocus, false)
>   window.addEventListener('online', handleOnline, false)

Is there any documentation on when/how to cleanup these event listeners ? When I added breakpoint in the code that removes the listener, that was never hit by debugger. Thanks.

> window.removeEventListener(.., ..)

**Comment by TkDodo** [comment, marked as answer]
I fixed this with v4.

**Comment by omar-oslo** [comment]
What about v 3.x ? Is it fixed there as well ?

**Reply by TkDodo** [reply]
No

**Reply by omar-oslo** [reply]
Thanks. Can you pls point to the commit which fixes these eventlistener cleanup in v4 ?

**Reply by TkDodo** [reply]
seems like I've actually fixed this for v3 as well:

- https://github.com/TanStack/query/pull/3092

**Reply by omar-oslo** [reply]
Thanks. Found the 3.x fix version(3.34.3) in this comment. https://github.com/TanStack/query/pull/3092#issuecomment-993482467
